File: backend/studio_artifacts.py
import json
import re
from typing import Any, Dict, List


def _strip_code_fence(value: str) -> str:
    text = value.strip()
    if text.startswith("```"):
        first_newline = text.find("\n")
        text = text[first_newline + 1:] if first_newline >= 0 else ""
    if text.rstrip().endswith("```"):
        text = text.rstrip()[:-3]
    return text.strip()


def _repair_json_backslashes(value: str) -> str:
    """Escape model-generated LaTeX commands and otherwise invalid JSON escapes."""
    repaired = re.sub(r"(?<!\\)\\(?=[A-Za-z]{2,})", r"\\\\", value)
    return re.sub(r'(?<!\\)\\(?!["\\/bfnrtu])', r"\\\\", repaired)


def _json_candidates(raw_response: str) -> List[str]:
    text = _strip_code_fence(raw_response)
    candidates = [text]
    # Prefer the complete item array over an individual object nested inside it.
    for opening, closing in (("[", "]"), ("{", "}")):
        start = text.find(opening)
        end = text.rfind(closing)
        if start >= 0 and end > start:
            candidates.append(text[start:end + 1])
    return list(dict.fromkeys(candidate for candidate in candidates if candidate))


def _load_model_json(raw_response: str) -> Any:
    last_error: Exception | None = None
    for candidate in _json_candidates(raw_response):
        for attempted in (candidate, _repair_json_backslashes(candidate)):
            try:
                return json.loads(attempted)
            except (json.JSONDecodeError, TypeError) as exc:
                last_error = exc
    raise ValueError("La respuesta no contiene JSON estructurado válido.") from last_error


def _recover_complete_objects(raw_response: str) -> List[Dict[str, Any]]:
    """Recover fully closed item objects from an otherwise truncated JSON response."""
    text = _strip_code_fence(raw_response)
    decoder = json.JSONDecoder()
    recovered: List[Dict[str, Any]] = []
    for position, character in enumerate(text):
        if character != "{":
            continue
        try:
            value, _ = decoder.raw_decode(text[position:])
        except json.JSONDecodeError:
            try:
                value, _ = decoder.raw_decode(_repair_json_backslashes(text[position:]))
            except json.JSONDecodeError:
                continue
        if isinstance(value, dict) and not any(
            key in value for key in ("data", "quiz", "flashcards", "items")
        ):
            recovered.append(value)
    return recovered


def _unwrap_items(value: Any, artifact_type: str) -> Any:
    if not isinstance(value, dict):
        return value
    for key in ("data", artifact_type, "items"):
        if key in value:
            return value[key]
    return value


def parse_studio_items(raw_response: str, artifact_type: str) -> List[Dict[str, Any]]:
    """Parse and validate quiz or flashcard items from an LLM response."""
    try:
        parsed = _unwrap_items(_load_model_json(raw_response), artifact_type)
    except ValueError:
        parsed = _recover_complete_objects(raw_response)
    if not isinstance(parsed, list) or not parsed:
        raise ValueError(f"La respuesta de {artifact_type} no contiene una lista de elementos.")

    normalized: List[Dict[str, Any]] = []
    for position, item in enumerate(parsed, start=1):
        if not isinstance(item, dict):
            raise ValueError(f"El elemento {position} de {artifact_type} no es un objeto.")

        if artifact_type == "flashcards":
            front = str(item.get("front", "")).strip()
            back = str(item.get("back", "")).strip()
            if not front or not back:
                raise ValueError(f"La tarjeta {position} no tiene anverso o reverso.")
            normalized.append({"id": position, "front": front, "back": back})
            continue

        if artifact_type == "quiz":
            question = str(item.get("question", "")).strip()
            options = item.get("options")
            explanation = str(item.get("explanation", "")).strip()
            correct_index = item.get("correct_index")
            if (
                not question
                or not isinstance(options, list)
                or len(options) < 2
                or not all(str(option).strip() for option in options)
                or not isinstance(correct_index, int)
                or not 0 <= correct_index < len(options)
            ):
                raise ValueError(f"La pregunta {position} tiene una estructura inválida.")
            normalized.append(
                {
                    "id": position,
                    "question": question,
                    "options": [str(option).strip() for option in options],
                    "correct_index": correct_index,
                    "explanation": explanation,
                }
            )
            continue

        raise ValueError(f"Tipo de artefacto estructurado desconocido: {artifact_type}")

    return normalized


def studio_output_schema(artifact_type: str, count: int) -> Dict[str, Any]:
    """Build the JSON schema sent to Ollama for constrained Studio output."""
    if artifact_type == "flashcards":
        item_schema: Dict[str, Any] = {
            "type": "object",
            "properties": {
                "id": {"type": "integer"},
                "front": {"type": "string"},
                "back": {"type": "string"},
            },
            "required": ["id", "front", "back"],
            "additionalProperties": False,
        }
    elif artifact_type == "quiz":
        item_schema = {
            "type": "object",
            "properties": {
                "id": {"type": "integer"},
                "question": {"type": "string"},
                "options": {
                    "type": "array",
                    "items": {"type": "string"},
                    "minItems": 2,
                },
                "correct_index": {"type": "integer"},
                "explanation": {"type": "string"},
            },
            "required": ["id", "question", "options", "correct_index", "explanation"],
            "additionalProperties": False,
        }
    else:
        raise ValueError(f"No existe un esquema estructurado para {artifact_type}.")

    return {
        "type": "object",
        "properties": {
            "data": {
                "type": "array",
                "items": item_schema,
                "minItems": count,
                "maxItems": count,
            }
        },
        "required": ["data"],
        "additionalProperties": False,
    }

File: frontend/src/utils/studioArtifacts.ts
import type { StudioResumePayload } from '../types';

export interface FlashcardItem {
  id: number;
  front: string;
  back: string;
}

export interface QuizItem {
  id: number;
  question: string;
  options: string[];
  correct_index: number;
  explanation: string;
}

type ArtifactRecord = StudioResumePayload & Record<string, unknown>;

function repairJsonBackslashes(value: string): string {
  return value
    .replace(/(?<!\\)\\(?=[A-Za-z]{2,})/g, '\\\\')
    .replace(/(?<!\\)\\(?!["\\/bfnrtu])/g, '\\\\');
}

function stripCodeFence(value: string): string {
  let text = value.trim();
  if (text.startsWith('```')) {
    const firstNewline = text.indexOf('\n');
    text = firstNewline >= 0 ? text.slice(firstNewline + 1) : '';
  }
  if (text.trimEnd().endsWith('```')) {
    text = text.trimEnd().slice(0, -3);
  }
  return text.trim();
}

function parseModelJson(value: string): unknown | null {
  const text = stripCodeFence(value);
  const candidates = [text];

  // Prefer the complete item array over an individual object nested inside it.
  for (const [opening, closing] of [['[', ']'], ['{', '}']] as const) {
    const start = text.indexOf(opening);
    const end = text.lastIndexOf(closing);
    if (start >= 0 && end > start) candidates.push(text.slice(start, end + 1));
  }

  for (const candidate of [...new Set(candidates)]) {
    for (const attempt of [candidate, repairJsonBackslashes(candidate)]) {
      try {
        return JSON.parse(attempt);
      } catch {
        // Try the next candidate or repaired representation.
      }
    }
  }
  return null;
}

function recoverCompleteObjects(value: string): Record<string, unknown>[] {
  const text = stripCodeFence(value);
  const starts: number[] = [];
  const recovered: Record<string, unknown>[] = [];
  let inString = false;
  let escaped = false;

  for (let index = 0; index < text.length; index += 1) {
    const character = text[index];
    if (inString) {
      if (escaped) escaped = false;
      else if (character === '\\') escaped = true;
      else if (character === '"') inString = false;
      continue;
    }
    if (character === '"') {
      inString = true;
    } else if (character === '{') {
      starts.push(index);
    } else if (character === '}' && starts.length > 0) {
      const start = starts.pop();
      if (start === undefined) continue;
      const fragment = text.slice(start, index + 1);
      for (const attempt of [fragment, repairJsonBackslashes(fragment)]) {
        try {
          const parsed = JSON.parse(attempt);
          const record = asRecord(parsed);
          if (record) recovered.push(record);
          break;
        } catch {
          // An outer object can still be incomplete; retain only valid inner objects.
        }
      }
    }
  }
  return recovered;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function unwrapArtifactItems(value: unknown, artifactType: 'flashcards' | 'quiz'): unknown {
  const record = asRecord(value);
  if (!record) return value;
  return record.data ?? record[artifactType] ?? record.items;
}

export function recoverFlashcards(value: unknown): FlashcardItem[] | null {
  let candidate = value;
  if (typeof candidate === 'string') {
    candidate = parseModelJson(candidate);
  }

  const record = asRecord(candidate);
  if (record) {
    const direct = unwrapArtifactItems(record, 'flashcards');
    if (direct !== undefined) candidate = direct;
    if (!Array.isArray(candidate) && typeof record.content === 'string') {
      candidate = unwrapArtifactItems(parseModelJson(record.content), 'flashcards');
    }
  }

  if (!Array.isArray(candidate) || candidate.length === 0) return null;

  const cards: FlashcardItem[] = [];
  for (const [index, item] of candidate.entries()) {
    const itemRecord = asRecord(item);
    const front = typeof itemRecord?.front === 'string' ? itemRecord.front.trim() : '';
    const back = typeof itemRecord?.back === 'string' ? itemRecord.back.trim() : '';
    if (!front || !back) return null;
    cards.push({ id: index + 1, front, back });
  }
  return cards;
}

export function recoverQuiz(value: unknown): QuizItem[] | null {
  let candidate = value;
  if (typeof candidate === 'string') {
    candidate = parseModelJson(candidate) ?? recoverCompleteObjects(candidate);
  }

  const record = asRecord(candidate);
  if (record) {
    const direct = unwrapArtifactItems(record, 'quiz');
    if (direct !== undefined) candidate = direct;
    if (!Array.isArray(candidate) && typeof record.content === 'string') {
      const parsedContent = parseModelJson(record.content) ?? recoverCompleteObjects(record.content);
      candidate = unwrapArtifactItems(parsedContent, 'quiz');
    }
  }
  if (!Array.isArray(candidate) || candidate.length === 0) return null;

  const questions: QuizItem[] = [];
  for (const item of candidate) {
    const itemRecord = asRecord(item);
    const question = typeof itemRecord?.question === 'string' ? itemRecord.question.trim() : '';
    const options = Array.isArray(itemRecord?.options)
      ? itemRecord.options.map(option => typeof option === 'string' ? option.trim() : '')
      : [];
    const correctIndex = itemRecord?.correct_index;
    const explanation = typeof itemRecord?.explanation === 'string' ? itemRecord.explanation.trim() : '';
    if (
      !question || options.length < 2 || options.some(option => !option)
      || !Number.isInteger(correctIndex) || Number(correctIndex) < 0 || Number(correctIndex) >= options.length
    ) continue;
    questions.push({
      id: questions.length + 1,
      question,
      options,
      correct_index: Number(correctIndex),
      explanation,
    });
  }
  return questions.length > 0 ? questions : null;
}

export function normalizeStudioArtifact(value: unknown, fallbackArtifactType = 'briefing'): ArtifactRecord {
  const parsedValue = typeof value === 'string' ? parseModelJson(value) : value;
  const record = asRecord(parsedValue) ?? { content: typeof value === 'string' ? value : '' };
  const artifactType = typeof record.artifact_type === 'string'
    ? record.artifact_type
    : fallbackArtifactType;

  if (artifactType === 'flashcards' || record.type === 'flashcards') {
    const cards = recoverFlashcards(record);
    if (cards) {
      return { ...record, type: 'flashcards', artifact_type: 'flashcards', data: cards };
    }
  }

  if (artifactType === 'quiz' || record.type === 'quiz') {
    const questions = recoverQuiz(record);
    if (questions) {
      return { ...record, type: 'quiz', artifact_type: 'quiz', data: questions };
    }
  }

  return {
    ...record,
    type: 'markdown',
    artifact_type: artifactType,
    content: typeof record.content === 'string'
      ? record.content
      : typeof value === 'string' ? value : '',
  };
}
